app/common/ACLRuleCollection.ts

```diff
@@ -183,10 +183,28 @@ export class ACLRuleCollection {
   }
 
   /**
-   * Update granular access from DocData.
+   * Update granular access from DocData. Try hard not to throw
+   * expections, reporting problems via this.ruleError. If we
+   * throw an exception, document recovery mode won't work.
    */
   public async update(docData: DocData, options: ReadAclOptions) {
-    const {ruleSets, userAttributes} = this._safeReadAclRules(docData, options);
+    try {
+      await this.updateWithExceptions(docData, options);
```

Comment:
Should this set `this.ruleError = undefined` at the start of the method? Checking because previous code did that at the start of `_safeReadAclRules`.